translations/translated/generate-translation-stats.lua

```diff
@@ -28,37 +28,53 @@ local stats = {}
 
 while line <= #lines do
   local currentLine = lines[line]
-  local lang = currentLine:match("mudlet_([a-z]+_[A-Z]+)%.qm")
+  local lang = currentLine:match("Updating 'mudlet_([a-z]+_[A-Z]+)%.qm'...")
   line = line + 1
   if lang then
-    local translated
+    local translated = 0
+    local finished = 0
+    local unfinished = 0
     repeat
       currentLine = lines[line]
-      translated = tonumber(currentLine:match("(%d+)"))
+      translated, finished, unfinished = currentLine:match("Generated (%d+) translation%(s%) %((%d+) finished and (%d+) unfinished%)")
+      if translated ~= nil and finished ~= nil and unfinished ~= nil then
+        translated = tonumber(translated)
+        finished = tonumber(finished)
+        unfinished = tonumber(unfinished)
+      end
       line = line + 1
     until translated ~= nil
     currentLine = lines[line]
-    local untranslated = tonumber(currentLine:match("(%d+)"))
-    line = line + 1
-
-    -- US English is always at 100% translation
-    if lang == "en_US" then
-      translated, untranslated = untranslated, 0
+    local untranslated = 0
+    if currentLine ~= nil then
+      -- The untranslated line will not be present if translation is at 100%!
+      untranslated = currentLine:match("Ignored (%d+) untranslated source text%(s%)")
+      if untranslated ~= nil then
+        untranslated = tonumber(untranslated)
+        -- There IS an "Ignored..." line so increment the line counter past it,
+        line = line + 1
+      else
+        -- reset from the nil returned value to 0 (as that how many untranslated
+        --  source text there are!)
+      end
+      -- Otherwise leave it unchanged for the next time around the outer loop
     end
 
     stats[#stats + 1] = {
       lang = lang,
       translated = translated,
       untranslated = untranslated,
+      finished = finished,
+      unfinished = unfinished,
       total = translated + untranslated,
       translatedpc = math.floor((100/(translated + untranslated)) * translated)
     }
   end
 end
 
-print("lang", "trnsl", "utrnsl", "total", "percentage")
+print("lang", "trnsl", "utrnsl", "finish", "unfin", "total", "percentage")
```

Comment:
Why do we need `unfin`? Looking at the logs, it's all 0's. How is that related to handling 100% complete translation?

Reply:
Those numbers might not be for the `en_US` case and, not being sure which ones are changed under whatever circumstances it isn't too unreasonable to grab all of them, is it?

Reply:
But why? They won't be used for anything as far as I know. Better to have simpler code, than more complicated code where you're not sure what's actually relevant or not 🤷‍♂️

Reply:
+1 YAGNI
Leave a comment if you fear forgetting the stat exists, but if it's not needed, then don't include it in code.

Reply:
I want to check, I suspect (but have to prove) that if more *numerous* (plaural) translations are added then they will show up there for the English (American) case - I further suspect that it might be that "untranslated"/"finished" and "unfinished" *might* relate to different numbers including:

* the total number of strings that are needed (which for languages that have plurals - pretty much all but Japan IIRC) that means **more than one** per source text when the `tr(...)` source has the third, `int`eger, argument.
* the number of strings that have an *unapproved* or *approved* translation (marked tags containing "unfinished" attribute in the `.ts` XML file) - counting each source text as a single item even if it needs more than one translation dependant on the plural form(s) for that locale.

As I say, I am not sure which of the figures is which, but will check and report what I find out.

Reply:
Ah, *I* didn't know about [**YAGNI**](https://en.wikipedia.org/wiki/You_aren't_gonna_need_it) - I had to look it up. As a counter - **not** that it applies in this case - can I propose the *term* **YAGNIBID** for the lexicon with a "... but I do" suffix? :rofl:

Reply:
For the record - when using Qt Linguist - as we will need to for the `mudlet_en_US.ts`/`mudlet_en_US.qm` case I *think* the column headings work out to mean:
- `lang` - guess what, the language code
- `trnsl` - the number of completed translations in the file - which is currently **`11`** for the en_US case
- `untrnsl` - the number of translations that do not have any texts as translations at all (and will be marked with a green query mark in Linguist)
- `finish` - the number of translations that have been marked as done in linguist (and have either a green{approved/done} or yellow {approved/done but has a warning} tick)
- `unfin` - the number of translations that have some text but are not marked as done in linguist (and thus have a yellow query{has entry but not marked as approved/done} mark or red exclamation {has entry but triggers a warning and not marked as approved/done} mark)
- `total` - the sum of `trnsl` and `untrnsl`

Whilst we will only see all or zero figures for the two additional columns for the other locales handled by CrowIn it is possible that other numbers will show for the American case if there are any changes to the wording in the ***numerous*** source texts to be translated...